File: namastack-outbox-examples/namastack-outbox-example-java/src/main/java/io/namastack/demo/DemoApplication.java
package io.namastack.demo;

import io.namastack.demo.customer.CustomerService;
import io.namastack.outbox.annotation.EnableOutbox;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.scheduling.annotation.EnableScheduling;

@EnableOutbox
@EnableScheduling
@SpringBootApplication
public class DemoApplication implements CommandLineRunner {

  private static final Logger logger = LoggerFactory.getLogger(DemoApplication.class);

  private final CustomerService service;

  public DemoApplication(CustomerService service) {
    this.service = service;
  }

  public static void main(String[] args) {
    SpringApplication.run(DemoApplication.class, args);
  }

  @Override
  public void run(@NotNull String... args) throws Exception {
    logger.info("=== Namastack Outbox Demo ===");

    logger.info("Register: John Wayne");
    var customer1 = service.register("John", "Wayne", "john.wayne@example.com");

    logger.info("Register: Macy Grey");
    var customer2 = service.register("Macy", "Grey", "macy.grey@example.com");

    logger.info("Waiting for processing...");
    Thread.sleep(2000);

    logger.info("Remove: {}", customer1.getId());
    service.remove(customer1.getId());

    logger.info("Remove: {}", customer2.getId());
    service.remove(customer2.getId());

    logger.info("=== Demo Complete ===");
  }
}

File: namastack-outbox-examples/namastack-outbox-basic-example/src/main/kotlin/io/namastack/demo/DemoApplication.kt
package io.namastack.demo

import io.namastack.demo.customer.CustomerService
import io.namastack.outbox.EnableOutbox
import org.slf4j.LoggerFactory
import org.springframework.boot.CommandLineRunner
import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.scheduling.annotation.EnableScheduling

/**
 * Namastack Outbox for Spring Boot Demo Application
 *
 * This application demonstrates the declarative @OutboxEvent approach for implementing
 * the transactional outbox pattern with Spring Boot and Spring Data JPA.
 *
 * How It Works:
 * 1. Events are marked with @OutboxEvent annotation
 * 2. Aggregates use @DomainEvents to collect events during state changes
 * 3. repository.save() triggers @DomainEvents publication
 * 4. OutboxEventMulticaster intercepts @OutboxEvent annotated events
 * 5. Events are automatically serialized and persisted to outbox
 * 6. DemoRecordProcessor polls and processes events asynchronously
 *
 * Key Components:
 * - Customer: Aggregate root with domain events
 * - CustomerService: Clean service without manual outbox code
 * - OutboxEventListener: Demonstrates optional event listener
 * - DemoRecordProcessor: Processes outbox records asynchronously
 *
 * @author Roland Beisel
 */
@EnableOutbox
@EnableScheduling
@SpringBootApplication
class DemoApplication(
    private val customerService: CustomerService,
) : CommandLineRunner {
    private val logger = LoggerFactory.getLogger(DemoApplication::class.java)

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            SpringApplication.run(DemoApplication::class.java, *args)
        }
    }

    override fun run(vararg args: String?) {
        logger.info("")
        logger.info("╔════════════════════════════════════════════════════════════╗")
        logger.info("║  Namastack Outbox for Spring Boot - Demo Application v0.3.0 ║")
        logger.info("║  Demonstrating @OutboxEvent + @DomainEvents Pattern         ║")
        logger.info("╚════════════════════════════════════════════════════════════╝")
        logger.info("")

        Thread.sleep(500)

        demonstrateCustomerLifecycle()

        logger.info("")
        logger.info("✨ Demo flow completed!")
        logger.info("📊 Events are being processed asynchronously by the Outbox Processor")
        logger.info("💡 Press Ctrl+C to stop the application")
        logger.info("")
    }

    private fun demonstrateCustomerLifecycle() {
        logger.info("─────────────────────────────────────────────────────────────")
        logger.info("Step 1: Creating Customers")
        logger.info("─────────────────────────────────────────────────────────────")

        val customer1 = customerService.register("John", "Smith")
        val customer2 = customerService.register("Alice", "Thompson")
        val customer3 = customerService.register("Bruno", "Bertolli")

        logger.info("✓ 3 customers registered")
        logger.info("")

        Thread.sleep(1000)

        logger.info("─────────────────────────────────────────────────────────────")
        logger.info("Step 2: Activating Customers")
        logger.info("─────────────────────────────────────────────────────────────")

        customerService.activate(customer1.id)
        customerService.activate(customer2.id)
        customerService.activate(customer3.id)

        logger.info("✓ 3 customers activated")
        logger.info("")

        Thread.sleep(1000)

        logger.info("─────────────────────────────────────────────────────────────")
        logger.info("Step 3: Deactivating Customers")
        logger.info("─────────────────────────────────────────────────────────────")

        customerService.deactivate(customer1.id)
        customerService.deactivate(customer2.id)
        customerService.deactivate(customer3.id)

        logger.info("✓ 3 customers deactivated")
        logger.info("")

        logger.info("─────────────────────────────────────────────────────────────")
        logger.info("Events Published to Outbox!")
        logger.info("─────────────────────────────────────────────────────────────")
        logger.info("Waiting for asynchronous processing...")
        logger.info("")
    }
}
